src/storage/storage_entry.rs

```diff
@@ -0,0 +1,371 @@
+{
+    let data = storage_entry::<S, T>(key.clone(), init);
+
+    StorageEntry::new_synced(key, data, cx)
+}
+
+#[allow(unused)]
+pub fn use_synced_storage_entry<S, T>(
+    cx: &ScopeState,
+    key: S::Key,
+    init: impl FnOnce() -> T,
+) -> UseStorageEntry<S, T>
+where
+    S: StorageBacking + StorageSubscriber<S>,
+    T: Serialize + DeserializeOwned + Clone + 'static,
+    S::Key: Clone,
+{
+    let state = use_signal(cx, || synced_storage_entry::<S, T>(key, init, cx));
+
+    if let Some(channel) = &state.read().channel {
+        use_listen_channel(cx, channel, move |message| async move { 
+            if let Ok(payload) = message {
+                if payload.key == state.read().key {
+                    state.write().update()
+                }
+            }
+        });
+    }
+    UseStorageEntry::new(state)
+}
+
+//  Start UseStorageEntry
+/// Storage that persists across application reloads
+pub struct UseStorageEntry<S: StorageBacking, T: Serialize + DeserializeOwned + Clone + 'static> {
+    inner: Signal<StorageEntry<S, T>>,
+}
+
+#[allow(unused)]
+impl<S: StorageBacking, T: Serialize + DeserializeOwned + Clone + 'static> UseStorageEntry<S, T> {
+    pub fn new(signal: Signal<StorageEntry<S, T>>) -> Self {
+        Self { inner: signal }
+    }
+
+    /// Returns a reference to the value
+    pub fn read(&self) -> StorageRef<S, T> {
+        StorageRef {
+            inner: self.inner.read(),
+        }
+    }
+
+    /// Returns a mutable reference to the value
+    pub fn write(&self) -> StorageRefMut<S, T> {
+        StorageRefMut {
+            inner: self.inner.write(),
+        }
+    }
+
+    /// Sets the value
+    pub fn set(&self, value: T) {
+        *self.write() = value;
+    }
+
+    /// Modifies the value
+    pub fn modify<F: FnOnce(&mut T)>(&self, f: F) {
+        f(&mut self.write());
+    }
+
+    pub fn save(&self) {
+        self.inner.write().save();
+    }
+}
+
+#[allow(unused)]
+impl<S: StorageBacking, T: Serialize + DeserializeOwned + Default + Clone + 'static> UseStorageEntry<S, T> {
+    /// Returns a clone of the value
+    pub fn get(&self) -> T {
+        self.read().clone()
+    }
+}
+
+impl<S: StorageBacking, T: Serialize + DeserializeOwned + Default + Display + Clone + 'static> Display
+    for UseStorageEntry<S, T>
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        (*self.read()).fmt(f)
+    }
+}
+
+impl<S: StorageBacking, T: Serialize + DeserializeOwned + Default + Clone + 'static> Deref for UseStorageEntry<S, T> {
+    type Target = Signal<StorageEntry<S, T>>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl<S: StorageBacking, T: Serialize + DeserializeOwned + Default + Clone + 'static> DerefMut for UseStorageEntry<S, T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.inner
+    }
+}
+
+pub struct StorageRef<'a, S: StorageBacking, T: Serialize + DeserializeOwned + Clone + 'static> {
```

Comment:
Can this just be Ref instead of it's own type? That would allow users to use methods like [map](https://doc.rust-lang.org/std/cell/struct.Ref.html#method.map)

Reply:
yeah i was thinking the same, will refactor

Reply:
Trying to map the `Write<Signal<StorageEntry<S,T>>>` to `Write<T>`, but the dioxus-signals `Write::map` function is acting funky. I am trying the following map:

```rust
Write::map(self.inner.write(), |entry: &mut StorageEntry<S, T>| &mut entry.data)
```

However, I'm seeing an error when trying to return this, where it's saying that the above is actually returning Write<'_, _, StorageEntry<S,T>>. I looked at the source code for this and I see that the Write struct is actually `Write<'a, T: 'static, I: 'static = T>`. I am not sure this map can ever work because of the restriction that I = T, meaning the SignalSubscriberDrop and the underlying RefMut must always be the same type...

Reply:
I've removed the StorageRef, etc. by swapping .read() for a Ref and removing write altogether. Instead of exposing a RefMut, I just expose .set() and .modify(). This gives better control over the save story, since now I can read whether a StorageBacking requires save on every change or can tolerate save only on drop.

Reply:
> Trying to map the `Write<Signal<StorageEntry<S,T>>>` to `Write<T>`, but the dioxus-signals `Write::map` function is acting funky. I am trying the following map:
> 
> ```rust
> Write::map(self.inner.write(), |entry: &mut StorageEntry<S, T>| &mut entry.data)
> ```
> 
> However, I'm seeing an error when trying to return this, where it's saying that the above is actually returning Write<'_, _, StorageEntry<S,T>>. I looked at the source code for this and I see that the Write struct is actually `Write<'a, T: 'static, I: 'static = T>`. I am not sure this map can ever work because of the restriction that I = T, meaning the SignalSubscriberDrop and the underlying RefMut must always be the same type...

I = T just means that if I isn't specified then it can be assumed to be equal to T. If you make the function return Write<'a, ORIGINAL_TYPE, NEW_TYPE> then it should work

Reply:
I was able to get this to work, by returning type `Write<'_, NEW_TYPE, ORIGINAL_TYPE>`, however I ran into an issue when dereferencing. I was able to resolve this by modifying the deref impl for Write to the following:

```rust
impl<'a, T: 'static, I: 'static> Deref for Write<'a, T, I> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.write
    }
}

impl<T, I> DerefMut for Write<'_, T, I> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.write
    }
}
```

Reply:
That is definitely an oversight. I generally rely on cargo to warn me if I am missing generics, but with default types both with and without the generic is valid. I wish there was a way to `#[warn(default_generic_impl)]`...

If you file a PR with your Dioxus fork I would be happy to merge it

Reply:
I've opened a PR for this: https://github.com/DioxusLabs/dioxus/pull/1549